### ISSUE TITLE
Toolchain isolation for scalac optoins

### ISSUE THREAD

**Issue by andyscott** [bazelbuild/rules_scala#432]
Downstream workspaces using a custom Scala toolchain can't use the `scala_test` rule with certain enable Scala compiler flags enabled.

It seems custom Scala toolchains will be used to compile the bootstrapping Scala code in `rules_scala` for running scala tests. In particular, flags to enable fatal warnings and warnings on unused imports cause `scala_test` to be unusable.

A short term fix is to clean up these bootstrapping files so that they work under strict Scala compiler flags. I took a look at a more thorough solution but I couldn't come up with anything quickly.

**Comment by ittaiz** [comment]
I’m not sure what you’re looking (toolchain isolation) for is an option in
Bazel due to source dependency.
How about opening an issue with bazel asking if it’s a relevant request?
Another option would be to build rules Scala with the strictest modes as to
support all users

**Comment by johnynek** [comment]
+1 for more strictness by default.

**Comment by andyscott** [comment]
@ittaiz I've adjusted the title a bit to maybe be a bit less misleading.
AFAIK toolchains operate globally.

**Comment by ittaiz** [comment]
I agree, that’s why I said it’s a bazel side issue if you want to change it.
Do you have any suggestions on how to insulate it other than building
rules_scala with the strictest defaults?

**Comment by ittaiz** [comment]
@andyscott I think we should either close this issue in favor of a bazel one or rename it to clarify this issue is solely about making rules_scala work with wfatal and the rest. WDYT?

**Comment by andyscott** [comment]
We could perhaps close this in favor of a more general toolchain issue.

Toolchains strike me as the standard way to get tools off the host environment. For Scala, this isn't needed, so I'd like to discuss the removal of toolchains for boostrapping Scala compilation.

**Comment by ittaiz** [comment]
More general in what sense?
Getting tools off of the host environment is one use case but toolchains in general belong to the world of configurability.
Configuring then scala SDK as well as the scalacopts are classic toolchain use cases.

Why should we remove toolchains from our scala code?

**Comment by dsilvasc** [comment]
You might be able to isolate settings for rules_scala with platform constraints: you could define a platform `@rules_scala//:bootstrap-platform` and set `restricted_to` to that platform in the targets defined in the rules_scala repo.

The consumers could have one scala toolchain for their application code (maybe stricter, maybe plus-one deps) and another scala toolchain to compile rules_scala (with whichever settings rules_scala requires).

**Comment by ittaiz** [comment]
interesting, doesn't that require support for using multi platform in a single invocation in bazel?
Bazel is very oriented towards everything being one mono repo together, the above scenario doesn't sound like it will work to be honest
